### PR TITLE
Add initial type specs

### LINKED ISSUE #57: function typespecs

### PR DESCRIPTION
**Pull request by princemaple**
Closes #57

### CODE REVIEW

File: lib/req/response.ex [hunk 1]
```diff
@@ -15,6 +15,8 @@ defmodule Req.Response do
       conflicts. Only accepts `t:atom/0` keys.
   """
 
+  @type t :: %__MODULE__{}
```
Comment (wojtekmach) [inline lib/req/response.ex:18]:
```suggestion
  @type t() :: %__MODULE__{
          status: non_neg_integer(),
          headers: [{binary(), binary()}],
          body: binary()
        }
```

File: lib/req.ex [hunk 2]
```diff
@@ -8,12 +8,28 @@ defmodule Req do
...
+
+  @type method :: :get | :post | :put | :delete
+
   @doc """
   Makes a GET request.
 
   See `request/3` for a list of supported options.
   """
   @doc api: :high_level
+  @spec get!(String.t(), options()) :: Req.Response.t()
```
[the remark this answers is not in the document]
Comment (princemaple) [reply lib/req.ex:32]: Hmm, why not leave it there for now and update it when you update the code? It'd be a breaking change anyway.

[the remark this answers is not in the document]
Comment (wojtekmach) [reply lib/req.ex:32]: I just haven't made my mind if I want to spec these or not, I have some ideas about options handling in the future and I don't want this change to hamper that.

**Comment by princemaple** [comment]
All updated.

File: lib/req/response.ex [hunk 3]
```diff
@@ -15,6 +15,12 @@ defmodule Req.Response do
       conflicts. Only accepts `t:atom/0` keys.
   """
 
+  @type t() :: %__MODULE__{
+          status: non_neg_integer(),
+          headers: [{binary(), binary()}],
+          body: binary()
```
Comment (wojtekmach) [inline lib/req/response.ex:21]:
```suggestion
          body: binary(),
          private: map()
```

File: lib/req.ex [hunk 4]
```diff
@@ -8,12 +8,15 @@ defmodule Req do
              |> String.split("<!-- MDOC !-->")
              |> Enum.fetch!(1)
 
+  @type method :: :get | :post | :put | :delete
```
Comment (wojtekmach) [inline lib/req.ex:11]:
```suggestion
  @type url() :: URI.t() | String.t()

  @type method() :: :get | :post | :put | :delete
```

File: lib/req.ex [hunk 5]
```diff
@@ -8,12 +8,15 @@ defmodule Req do
...
+
   @doc """
   Makes a GET request.
 
   See `request/3` for a list of supported options.
   """
   @doc api: :high_level
+  @spec get!(String.t(), keyword()) :: Req.Response.t()
```
Comment (wojtekmach) [inline lib/req.ex:19]:
```suggestion
  @spec get!(url(), keyword()) :: Req.Response.t()
```

File: lib/req.ex [hunk 6]
```diff
@@ -24,6 +27,7 @@ defmodule Req do
   See `request/3` for a list of supported options.
   """
   @doc api: :high_level
+  @spec post!(String.t(), term(), keyword()) :: Req.Response.t()
```
Comment (wojtekmach) [inline lib/req.ex:30]:
```suggestion
  @spec post!(url(), body :: term(), keyword()) :: Req.Response.t()
```

File: lib/req.ex [hunk 7]
```diff
@@ -35,6 +39,7 @@ defmodule Req do
   See `request/3` for a list of supported options.
   """
   @doc api: :high_level
+  @spec put!(String.t(), term(), keyword()) :: Req.Response.t()
```
Comment (wojtekmach) [inline lib/req.ex:42]:
```suggestion
  @spec put!(url(), body :: term(), keyword()) :: Req.Response.t()
```

File: lib/req.ex [hunk 8]
```diff
@@ -46,6 +51,7 @@ defmodule Req do
   See `request/3` for a list of supported options.
   """
   @doc api: :high_level
+  @spec delete!(String.t(), keyword()) :: Req.Response.t()
```
Comment (wojtekmach) [inline lib/req.ex:54]:
```suggestion
  @spec delete!(url(), keyword()) :: Req.Response.t()
```

File: lib/req.ex [hunk 9]
```diff
@@ -71,6 +77,7 @@ defmodule Req do
   The `options` are merged with default options set with `default_options/1`.
   """
   @doc api: :high_level
+  @spec request(method(), String.t(), keyword()) :: {:ok, Req.Response.t()} | {:error, Exception.t()}
```
Comment (wojtekmach) [inline lib/req.ex]:
```suggestion
  @spec request(method(), url(), keyword()) :: {:ok, Req.Response.t()} | {:error, Exception.t()}
```

File: lib/req.ex [hunk 10]
```diff
@@ -86,6 +93,7 @@ defmodule Req do
   See `request/3` for more information.
   """
   @doc api: :high_level
+  @spec request!(method(), String.t(), keyword()) :: Req.Response.t()
```
Comment (wojtekmach) [inline lib/req.ex:97]:
```suggestion
  @spec request!(method(), url(), keyword()) :: Req.Response.t()
```

File: lib/req.ex [hunk 11]
```diff
@@ -71,6 +77,8 @@ defmodule Req do
   The `options` are merged with default options set with `default_options/1`.
   """
   @doc api: :high_level
+  @spec request(method(), String.t(), keyword()) ::
```
Comment (wojtekmach) [inline lib/req.ex:80]:
```suggestion
  @spec request(method(), url(), keyword()) ::
```

**Comment by wojtekmach** [comment]
Thank you!

**Pull request merged**